### DISCUSSION TITLE
Correct way to represpen close state of controlled Navigation Menu

### DISCUSSION THREAD

**Discussion by QzCurious** [radix-ui/primitives#2124] [Help] [answered]
[Sample code](https://codesandbox.io/p/sandbox/radix-correct-way-to-represpen-close-state-of-controlled-navigation-menu-tlo0cn)

## Not work

The type for `value` of `NavigationMenu.Root` is `value?: string`. So it is valid:

```tsx
const [openingMenu, setOpeningMenu] = React.useState<string>()
    // ^ string | undefined
return <NavigationMenu.Root
  value={openingMenu}
  onValueChange={setOpeningMenu}
/>
```

But since it is **controlled**, for react, `value` should not be `undefined`. It also makes it unclear what should `NavigationMenu.Root` do for `setOpeningMenu(undefined)`.

## Work

I'm currently mark close state as empty string and it seems to work.

```ts
const [openingMenu, setOpeningMenu] = React.useState('')
// close NavigationMenuby
setOpeningMenu('')
```

## My Question

1. The correct way to controll Navigation Menu
2. If it's good to include an example or some statement to the doc
3. Whether `NavigationMenu.Root` should be close if the controlled value is set to `undefined`

## Some clips for the sample code

https://user-images.githubusercontent.com/38932402/235632159-44a4ef54-adc1-46e1-a1f6-693958d41233.mp4

https://user-images.githubusercontent.com/38932402/235632165-c6084b12-7556-4b78-8621-72d151ebf8d4.mp4

https://user-images.githubusercontent.com/38932402/235632168-54489168-a112-4293-9c83-e6eb098c29ce.mp4

**Comment by benoitgrelard** [comment, marked as answer]
> But since it is controlled, for react, value should not be undefined. It also makes it unclear what should NavigationMenu.Root do for setOpeningMenu(undefined).

Yes that's correct. Controlling it and setting it back to `undefined` is essentially trying to set it back to "uncontrolled" so should be avoided. The same is true for any component in React (ie. the same is true with native `input` for example and React will actually warn you if you do this).

**Reply by QzCurious** [reply]
Thanks.

For controlled `NavigationMenu`, a `''` (or any string that more meaningful) as **reset value** for close state is the way to go.
